bd-client-stats/src/stats.rs

```diff
@@ -9,347 +9,127 @@
-    Ok(())
-  }
-
-  async fn read_compressed_protobuf<T: protobuf::Message>(
-    &self,
-    path: impl AsRef<Path> + Send,
-  ) -> anyhow::Result<T> {
-    let compressed_bytes = self.read_file(path).await?;
-    Ok(read_compressed_protobuf(&compressed_bytes)?)
-  }
+pub trait Ticker: Send + Sync {
+  async fn tick(&mut self);
 }
 
 //
-// RealSerializedFileSystem
+// RuntimeWatchTicker
 //
 
-/// The real filesystem implementation which delegates to `tokio::fs`, joining the relative paths
-/// provided in the calls with the SDK directory.
-pub struct RealSerializedFileSystem {
-  directory: PathBuf,
-  semaphore: Semaphore,
+pub struct RuntimeWatchTicker {
+  receiver: watch::Receiver<Duration>,
+  interval: Option<tokio::time::Interval>,
 }
 
-impl RealSerializedFileSystem {
+impl RuntimeWatchTicker {
   #[must_use]
-  pub fn new(directory: PathBuf) -> Self {
+  pub const fn new(receiver: watch::Receiver<Duration>) -> Self {
     Self {
-      directory,
-      semaphore: Semaphore::new(1),
+      receiver,
+      interval: None,
     }
   }
 }
 
 #[async_trait]
-impl SerializedFileSystem for RealSerializedFileSystem {
-  async fn read_file(&self, path: impl AsRef<Path> + Send) -> anyhow::Result<Vec<u8>> {
-    let _permit = self.semaphore.acquire().await;
-    Ok(tokio::fs::read(self.directory.join(path)).await?)
-  }
-
-  async fn write_file(
-    &self,
-    path: impl AsRef<Path> + Send,
-    data: impl AsRef<[u8]> + Send,
-  ) -> anyhow::Result<()> {
-    let _permit = self.semaphore.acquire().await;
-    Ok(tokio::fs::write(self.directory.join(path), data).await?)
-  }
-
-  async fn delete_file(&self, path: impl AsRef<Path> + Send) {
-    let _permit = self.semaphore.acquire().await;
-    handle_unexpected(
-      tokio::fs::remove_file(self.directory.join(path)).await,
-      "delete file",
-    );
-  }
-
-  async fn exists(&self, path: impl AsRef<Path> + Send) -> bool {
-    let _permit = self.semaphore.acquire().await;
-    tokio::fs::metadata(self.directory.join(path)).await.is_ok()
-  }
-}
-
-//
-// StatsUploader
-//
-
-/// Responsible for periodically preparing an upload request from the aggregated stats file to
-/// a "pending upload" file and periodically retrying this upload.
-pub struct Uploader<F: SerializedFileSystem> {
-  shutdown: ComponentShutdown,
-  upload_interval_flag: DurationWatch<UploadStatFlushIntervalFlag>,
-  data_flush_tx: Sender<DataUpload>,
-  fs: Arc<F>,
-}
-
-impl<F: SerializedFileSystem> Uploader<F> {
-  pub const fn new(
-    shutdown: ComponentShutdown,
-    upload_interval_flag: DurationWatch<UploadStatFlushIntervalFlag>,
-    data_flush_tx: Sender<DataUpload>,
-    fs: Arc<F>,
-  ) -> Self {
-    Self {
-      shutdown,
-      upload_interval_flag,
-      data_flush_tx,
-      fs,
+impl Ticker for RuntimeWatchTicker {
+  async fn tick(&mut self) {
+    if self.interval.is_none() {
+      self.interval = Some(self.receiver.borrow_and_update().jittered_interval_at());
```

Comment:
So does this mean that we are jittering all the intervals? Which would mean that if we set the interval to say 2m the average interval is more like 1m (depending on the jitter distribution)? If anything might be nice to have this as part of the doc string on the runtime flag

Reply:
No this just jitters the start of the interval. I will add a comment.